### DISCUSSION TITLE
How to make wrapper for input?

### DISCUSSION THREAD

**Discussion by 7iomka** [mantinedev/mantine#2015] [Q&A] [answered]
Hi.
By design I have all inputs looks like contained-floating.
<img width="417" alt="image" src="https://user-images.githubusercontent.com/18473137/183066492-01e78e48-d801-444e-b3a3-3bb401d81a2b.png">
Also I need to make some adapters to use mantine inputs with my form library.

The main problem is to how to make this wrapper to keep support both ways of usage - controlled & uncontrolled?

https://codesandbox.io/s/optimistic-davinci-kb9vse?file=/src/shared/ui/text-input.tsx

I tried to make a wrapper that needs information about the focus state and value, but as you can see, it looks messy and incomplete - if you enter data into the input and remove the focus, the label will return to its place.
Kindly tell me please how to design such wrappers over inputs?
Thanks a lot.

**Comment by rtivital** [comment]
Create a wrapper component that will hadle this logic, see Mantine UI as reference – https://ui.mantine.dev/category/inputs#floating-label-input
Currently, there are no other options.

**Reply by 7iomka** [reply]
> 

https://codesandbox.io/s/hungry-sutherland-ioq5i8?file=/src/shared/ui/text-input.tsx
That is correct way to have wrapper components for reuse cases which keeping both ways of use - controlled & uncontrolled?

**Reply by rtivital** [reply]
Yes, it seems to be correct

**Comment by 7iomka** [comment, marked as answer]
@rtivital Hi, I maked styles for how I want to see my inputs by design.
But now I have a problem: I need a type of input that is not present in your library - PhoneInput
I read the docs, so seems I need to create a [custom component](https://mantine.dev/core/input/#custom-component)
But for me is hard to apply create styles and logic for this case, how I made in [wrapper over text-input](https://codesandbox.io/s/hungry-sutherland-ioq5i8?file=/src/shared/ui/text-input.tsx)
The documentation in this regard does not tell how to do this.
Could you help me with a hypothetical example?
I put my questions in the partial of draft of code, many thanks.

```tsx
const useContainedFloatingStyles = createStyles(
  (theme, { floating }: { floating: boolean }) => ({
    root: {
      position: "relative"
    },

    input: {
      height: 48,
      paddingLeft: theme.spacing.xl,
      paddingRight: theme.spacing.xl,
      paddingTop: theme.spacing.sm,
      fontWeight: 700,
      "&::placeholder": {
        transition: "color 150ms ease",
        color: "transparent" // disable placeholders
      }
    },

    label: {
      position: "absolute",
      zIndex: 2,
      lineHeight: 1,
      top: theme.spacing.md,
      left: theme.spacing.xl,
      pointerEvents: "none",
      color: floating
        ? theme.colorScheme === "dark"
          ? theme.white
          : theme.fn.rgba(theme.black, 0.3)
        : theme.colorScheme === "dark"
        ? theme.colors.dark[3]
        : theme.fn.rgba(theme.black, 0.5),
      transition:
        "transform 150ms ease, color 150ms ease, font-size 150ms ease",
      transform: floating ? `translate(0, -${theme.spacing.md / 2}px)` : "none",
      fontSize: floating ? theme.fontSizes.xs : theme.fontSizes.md,
      fontWeight: 400
    },
    required: {
      transition: "opacity 150ms ease",
      opacity: floating ? 1 : 0
    }
  })
);

export type PhoneInputStylesNames = InputStylesNames | InputWrapperStylesNames;

interface PhoneInputProps
  extends DefaultProps<PhoneInputStylesNames>,
    InputSharedProps,
    InputWrapperBaseProps,
    Omit<React.ComponentPropsWithoutRef<'input'>, 'size'> {
  /* outer ref as an alternative for forwardRef */
  inputRef?: RefObject<HTMLInputElement>;
  /** Input element type */
  type?: React.HTMLInputTypeAttribute;

  /** Props passed to root element (InputWrapper component) */
  wrapperProps?: Record<string, any>;

  /** Input size */
  size?: MantineSize;

  __staticSelector?: string;
}

const defaultProps: Partial<PhoneInputProps> = {
  type: 'text',
  size: 'md',
  __staticSelector: 'PhoneInput',
};

// component body

const { inputProps, wrapperProps, inputRef, ...rest } = useInputProps(
  'PhoneInput',
  defaultProps,
  props,
);

const { classes } = useContainedFloatingStyles({
      floating: HOW TO CALCULATE? WHERE IS THE VALUE?,
      // floating: String(value).trim().length !== 0 || focused,
});

// Where to pass classes?

return (
<Input.Wrapper  {...wrapperProps}>
      <Input
        component={InputMask}
        mask="+7 (999) 999-99-99"
        ref={inputRef}
        {...inputProps}
        {...rest}
      />
</Input.Wrapper>

)

```

**Reply by rtivital** [reply]
Use Input.Base component and apply styles like in any other input – https://mantine.dev/core/input/#inputbase-component
